### PR TITLE
[Examples] Add pure gpu example of GraphSAGE link prediction

### PR DESCRIPTION
**Pull request by Kh4L**
## Description
Add an example of running GraphSAGE link prediction task from GPU memory. This was available before new dataloader, so this re-adds an example of that. The majority of this is the link_pred.py script, but the modified to store the graph on the GPU, and then several optimizations for inference.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] Code is well-documented
- [X] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change

### CODE REVIEW

File: examples/pytorch/graphsage/link_pred.py [hunk 1]
```diff
@@ -11,6 +12,18 @@
...
 
+class ArgParser(argparse.ArgumentParser):
+    def __init__(self):
+        super(ArgParser, self).__init__()
+        self.add_argument('--pure-gpu', action='store_true',
+                          help='mix CPU and GPU training')
+
+    def parse_args(self):
+        args = super().parse_args()
+        return args
```
Comment (BarclayII) [inline examples/pytorch/graphsage/link_pred.py:23]: Why are we subclassing `ArgumentParser`?  I think what people usually do is
```python
parser = argparse.ArgumentParser()
parser.add_argument(...)
args = parser.parse_args()
```

Three lines of code.

**Review by BarclayII — APPROVED** [review]
LGTM except for the ArgParser usage.

File: examples/pytorch/graphsage/link_pred.py (hunk 1 above)
Comment (Kh4L) [reply examples/pytorch/graphsage/link_pred.py]: Done

File: examples/pytorch/graphsage/link_pred.py [hunk 2]
```diff
@@ -11,6 +12,11 @@
 # (This is a long-standing issue)
 from ogb.linkproppred import DglLinkPropPredDataset
 
+parser = argparse.ArgumentParser()
+parser.add_argument('--pure-gpu', action='store_true',
+                    help='mix CPU and GPU training')
```
Comment (jermainewang) [inline examples/pytorch/graphsage/link_pred.py:17]:
```suggestion
                    help='Perform both sampling and training on GPU.')
```

File: examples/pytorch/graphsage/advanced/README.md [hunk 3]
```diff
@@ -3,10 +3,16 @@ More Examples for Training GraphSAGE
...
 
+#### Node classification task
 ```bash
 python3 pure_gpu_node_classification.py
 ```
 
+#### Link prediction task
+```
+python3 pure_gpu_link_pred.py
+```
```
Comment (jermainewang) [inline examples/pytorch/graphsage/advanced/README.md:14]: Remove

**Comment by jermainewang** [comment]
Hi @Kh4L , I like your change! It's very clean. Would you please also make changes to our node classification examples too? Thanks.

File: examples/pytorch/graphsage/link_pred.py (hunk 2 above)
Comment (Kh4L) [reply examples/pytorch/graphsage/link_pred.py:17]: Something went wrong with the copy-pastes, fixed!

**Comment by Kh4L** [comment]
Hi @jermainewang , thanks I am glad that you like it! 
Should I add it in this PR or open another one?

**Comment by jermainewang** [comment]
Adding it in this PR will be great!

**Comment by Kh4L** [comment]
> Adding it in this PR will be great!

Done!

File: examples/pytorch/graphsage/node_classification.py [hunk 4]
```diff
@@ -31,29 +36,32 @@ def forward(self, blocks, x):
...
+
         if buffer_device is None:
             buffer_device = device
 
         for l, layer in enumerate(self.layers):
-            y = torch.zeros(
+            y = torch.empty(
                 g.num_nodes(), self.n_hidden if l != len(self.layers) - 1 else self.n_classes,
-                device=buffer_device)
+                device=buffer_device, pin_memory=True)
```
Comment (jermainewang) [inline examples/pytorch/graphsage/node_classification.py:52]: In the `link_pred` example, `pin_memory` is set to be `args.pure_gpu`. Why different?

File: examples/pytorch/graphsage/link_pred.py [hunk 5]
```diff
@@ -69,23 +75,26 @@ def forward(self, pair_graph, neg_pair_graph, blocks, x):
...
 
         for l, layer in enumerate(self.layers):
-            y = torch.zeros(g.num_nodes(), self.n_hidden, device=buffer_device)
+            y = torch.zeros(g.num_nodes(), self.n_hidden, device=buffer_device,
+                            pin_memory=args.pure_gpu)
+            feat = feat.to(device)
+
             for input_nodes, output_nodes, blocks in tqdm.tqdm(dataloader):
-                x = blocks[0].srcdata['h']
+                x = feat[input_nodes]
```
Comment (jermainewang) [inline examples/pytorch/graphsage/link_pred.py:92]: @BarclayII I thought your previous implementation used prefetching for intermediate results. Is it critical? Does this change break that?

Comment (BarclayII) [reply examples/pytorch/graphsage/link_pred.py:92]: I think it has performance impacts.  Usability-wise it should be OK.

**Pull request merged**